qiskit/aqua/algorithms/amplitude_amplifiers/grover.py

```diff
@@ -111,192 +128,274 @@ def __init__(self,
+        else:
+            lam = 1.34
+
+        if num_iterations is not None:
+            validate_min('num_iterations', num_iterations, 1)
+            warnings.warn('The num_iterations argument is deprecated as of 0.8.0, and will be '
+                          'removed no earlier than 3 months after the release date. '
+                          'If you want to use the num_iterations argument '
+                          'you should use the iterations argument instead and pass an integer '
+                          'for the number of iterations.',
+                          DeprecationWarning, stacklevel=2)
 
         self._oracle = oracle
-        self._mct_mode = mct_mode
-        self._init_state = \
-            init_state if init_state else Custom(len(oracle.variable_register), state='uniform')
-        self._init_state_circuit = \
-            self._init_state.construct_circuit(mode='circuit', register=oracle.variable_register)
-        self._init_state_circuit_inverse = self._init_state_circuit.inverse()
-
-        self._diffusion_circuit = self._construct_diffusion_circuit()
-        self._max_num_iterations = np.ceil(2 ** (len(oracle.variable_register) / 2))
+        # Construct GroverOperator circuit
+        if grover_operator is not None:
+            self._grover_operator = grover_operator
+        else:
+            # check the type of state_preparation
+            if isinstance(state_preparation, InitialState):
+                warnings.warn('Passing an InitialState component is deprecated as of 0.8.0, and '
+                              'will be removed no earlier than 3 months after the release date. '
+                              'You should pass a QuantumCircuit instead.',
+                              DeprecationWarning, stacklevel=2)
+                if isinstance(oracle, Oracle):
+                    state_preparation = init_state.construct_circuit(
+                        mode='circuit', register=oracle.variable_register
+                        )
+                else:
+                    raise TypeError('If init_state is of type InitialState, oracle must be of type '
+                                    'Oracle')
+            elif not (isinstance(state_preparation, QuantumCircuit) or state_preparation is None):
+                raise TypeError('Unsupported type "{}" of state_preparation'.format(
+                    type(state_preparation)))
+
+            # check to oracle type and if necessary convert the deprecated Oracle component to
+            # a circuit
+            reflection_qubits = None
+            if isinstance(oracle, Oracle):
+                if not callable(getattr(oracle, "evaluate_classically", None)):
+                    raise AquaError(
+                        'Missing the evaluate_classically() method \
+                            from the provided oracle instance.'
+                    )
+                warnings.warn('Passing an qiskit.aqua.components.oracles.Oracle object is '
```

Comment:
Why are we deprecating Oracle here - the TruthTable and LogicalExpression Oracles do not have plain circuit versions yet - do they? These components are not deprecated